controller/konnect/konnectextension_controller_utils.go

```diff
@@ -33,39 +29,95 @@ import (
+		cpNN := client.ObjectKey{
+			Name:      ext.Spec.Konnect.ControlPlane.Ref.KonnectNamespacedRef.Name,
+			Namespace: ext.Namespace,
+		}
+		kgcp := &konnectv1alpha1.KonnectGatewayControlPlane{}
+		// Set the controlPlaneRefValidCond to false in case the KonnectGatewayControlPlane is not found.
+		if err := r.Get(ctx, cpNN, kgcp); err != nil {
+			if k8serrors.IsNotFound(err) {
+				getFromK8sErr = err
+			} else {
+				return nil, ctrl.Result{}, err
+			}
+		}
+		cp = kgcp
+	case commonv1alpha1.ControlPlaneRefKonnectID:
+		kgcpList := &konnectv1alpha1.KonnectGatewayControlPlaneList{}
+		if err := r.List(ctx, kgcpList, client.InNamespace(ext.Namespace)); err != nil {
+			return nil, ctrl.Result{}, err
+		}
+		kgcps := lo.Filter(kgcpList.Items, func(kgcp konnectv1alpha1.KonnectGatewayControlPlane, _ int) bool {
+			return kgcp.Status.ID == *ext.Spec.Konnect.ControlPlane.Ref.KonnectID
+		})
+		switch l := len(kgcps); l {
+		case 0:
+			getFromK8sErr = k8serrors.NewNotFound(
+				konnectv1alpha1.Resource("KonnectGatewayControlPlane"),
+				fmt.Sprintf("with KonnectID %s in namespace %s",
+					*ext.Spec.Konnect.ControlPlane.Ref.KonnectID, ext.Namespace,
+				),
+			)
+		case 1:
+			cp = &kgcps[0]
+		default:
+			return nil, ctrl.Result{}, fmt.Errorf(
+				"multiple KonnectGatewayControlPlanes found %d with the same KonnectID %s",
+				l, *ext.Spec.Konnect.ControlPlane.Ref.KonnectID,
+			)
+		}
+	}
+
 	controlPlaneRefValidCond := metav1.Condition{
 		Type:    konnectv1alpha1.ControlPlaneRefValidConditionType,
 		Status:  metav1.ConditionTrue,
 		Reason:  konnectv1alpha1.ControlPlaneRefReasonValid,
 		Message: "ControlPlaneRef is valid",
 	}
 
-	konnectCPID, res, err := r.getKonnectControlPlaneID(ctx, ext, dependingConditions...)
-	if err != nil || !res.IsZero() {
-		return nil, res, err
+	// Check if the KonnectGatewayControlPlane has been found.
+	if getFromK8sErr != nil {
+		controlPlaneRefValidCond.Status = metav1.ConditionFalse
+		controlPlaneRefValidCond.Reason = konnectv1alpha1.ControlPlaneRefReasonInvalid
+		controlPlaneRefValidCond.Message = getFromK8sErr.Error()
+		if res, _, errPatch := patch.StatusWithConditions(
+			ctx,
+			r.Client,
+			&ext,
+			append(dependingConditions, controlPlaneRefValidCond)...,
+		); errPatch != nil || !res.IsZero() {
+			return nil, res, errPatch
+		}
+		return nil, ctrl.Result{}, getFromK8sErr
 	}
 
-	// get the Konnect Control Plane from Konnect
-	konnectCP, err := ops.GetControlPlaneByID(ctx, sdk, konnectCPID)
-	// set the controlPlaneRefValidCond to false in case the Control Plane is not found in Konnect
-	if err != nil {
+	// Set the controlPlaneRefValidCond to false in case the KonnectGatewayControlPlane is not programmed yet.
+	if !lo.ContainsBy(cp.Status.Conditions, func(cond metav1.Condition) bool {
+		return cond.Type == konnectv1alpha1.KonnectEntityProgrammedConditionType &&
+			cond.Status == metav1.ConditionTrue
+	}) {
```

Comment:
For this we have `HasConditionTrue` and `HasConditionFalse` https://github.com/Kong/gateway-operator/blob/f769bfc62d6945f526b87abe668a989428842e14/pkg/utils/kubernetes/status.go#L85-L93